Improve the error message for deprecated preferences

alalazo opened the pull request:
This PR improves the warning for deprecated preferences, and adds a configuration audit to get files:lines details of the issues. The result is the following:

![Screenshot from 2023-11-16 17-59-17](https://github.com/spack/spack/assets/4199709/6889311b-dfbb-49c7-afdf-ea53389040d1)

The audit is to avoid more complex analysis every time the config file is loaded.

alalazo committed:
Improve the error message for deprecated preferences

haampie commented:
Maybe it should say that these preferences do not affect concretization?

Otherwise it seems that spack 0.22.dev actively ignores them and is actually breaking behavior

marcmengel reviewed (approved):
It's a definite improvement.  I'll take it.   
(I still think line/file info on the warning would be a Good Thing, but that is probably buried in the yaml parser somewhere... )

tldahlgren commented on lib/spack/spack/schema/packages.py:
```diff
@@ -162,10 +164,12 @@
-                    "v0.22.\n\n\tThese preferences will be ignored by Spack. You "
-                    "can set them only in the 'all' section of the same file.\n",
+                    "message": "setting 'compiler:', 'target:' or 'provider:' preferences in "
+                    "a package specific section of packages.yaml is deprecated, and will be "
```
A nit:

```suggestion
                    "a package-specific section of packages.yaml is deprecated, and will be "
```

tldahlgren commented on lib/spack/spack/schema/packages.py:
```diff
@@ -69,6 +69,8 @@
     "patternProperties": {r"\w+": {}},
 }
 
+REQUIREMENT_URL = "https://spack.readthedocs.io/en/latest/build_settings.html#package-requirements"
```
Want to confirm that you are pointing people to the Concretization Settings and not package requirements at https://spack.readthedocs.io/en/latest/packages_yaml.html#package-requirements, correct?

alalazo replied:
Thanks, I was pointing them at package requirements, but the link changed yesterday :wink:

alalazo replied:
Updated to new URL.

tldahlgren reviewed (commented):
Just a nit and question.

alalazo committed:
Add audit to check config files for deprecated attributes

The audit shows also the files:lines where the error occurs.

Also, address review comments.

alalazo committed:
Improve wording

Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

alalazo commented:
@marcmengel I pushed an improvement. This is what the audit shows on the config you gave me yesterday:
![Screenshot from 2023-11-16 18-20-54](https://github.com/spack/spack/assets/4199709/b677345d-afc6-49fb-b9f7-99ed2bac0ba6)

tldahlgren reviewed (approved):
LGTM.  Thanks.
